### PR TITLE
Make it easier for downstream to modify settings

### LINKED ISSUE #1159: make it easier to customize default host, password, port in the url via code
Is there anyway to pre-load / load the information (host, port, password) into the vnc.html page via PHP ?

This is possible with the vnc_light.html file but seems quite tricky with the vnc.html as it loads all the information from ui.js 

Thanks!

### PR DESCRIPTION
**Pull request by CendioOssman**
Expose a simple and stable API to override default settings, and force settings that users shouldn't be able to change.

Fixes #1159.

### CODE REVIEW

**Comment by CendioOssman** [comment]
I just realised this isn't going to work with `use_require.js`. So either that needs to be fixed somehow, or we have to wait until we can drop Internet Explorer and require module support.

Putting this on hold for now.

**Comment by CendioOssman** [comment]
So #1477 is now up, so we just need to wait for that.

**Comment by tculp** [comment]
Bump because this would be very useful and #1477 is now merged

**Comment by CendioOssman** [comment]
It's still on my todo, but I'm unsure when I'll have time to finish this.

File: vnc.html [hunk 1]
```diff
@@ -49,11 +49,40 @@
...
+            if (!response.ok) {
+                throw Error("" + response.status + " " + response.statusText);
+            }
+
+            mandatory = await response.json();
+        } catch (err) {
+            Log.Error("Couldn't fetch mandatory.json: " + err);
+        }
+
+        // You can also override any defaults you need here:
```
Comment (samhed) [inline vnc.html:85]: Neat!

Could we add a comment about how to find names of settings?

File: defaults.json [hunk 2]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (samhed) [inline defaults.json:1]: Should these include a commented-out example?

**Review by samhed — CHANGES_REQUESTED** [review]
Works great, this is a nice improvement!

I tested both setting stuff in vnc.html and in the defaults.json/mandatory.json, things work as expected as long as you clear the web storage inbetween, or use a private-browsing tab.

File: app/ui.js [hunk 3]
```diff
@@ -175,8 +187,6 @@ const UI = {
         UI.initSetting('repeaterID', '');
         UI.initSetting('reconnect', false);
         UI.initSetting('reconnect_delay', 5000);
-
-        UI.setupSettingLabels();
```
[the remark this answers is not in the document]
Comment (CendioOssman) [reply app/ui.js:175]: The new code disables the UI for mandatory settings. The mapping between controls and labels need to be set up before we go through the settings for the labels to also be disabled properly.

File: app/ui.js [hunk 4]
```diff
@@ -24,6 +24,8 @@ const LINGUAS = ["cs", "de", "el", "es", "fr", "it", "ja", "ko", "nl", "pl", "pt
 
 const UI = {
 
+    userSettings: {},
```
[the remark this answers is not in the document]
Comment (CendioOssman) [reply app/ui.js:27]: Sure, I understand the confusion. I think "custom" might be even better descriptor.

And the target audience in my mind are mainly sysadmins, not developers.

File: vnc.html (hunk 1 above)
Comment (CendioOssman) [reply vnc.html:85]: Sure. Don't know how to best refer to it, though. Just `"docs/EMBEDDING.md"`?

File: defaults.json (hunk 2 above)
Comment (CendioOssman) [reply defaults.json:1]: JSON doesn't support comments 🤬

**Comment by CendioOssman** [comment]
Hmm... It looks like the tests are broken on `master`. Something with the new H.264 code. Open a new issue and ping the author(s) of that code?

File: vnc.html (hunk 1 above)
Comment (samhed) [reply vnc.html:85]: Sounds good!

**Review by samhed — APPROVED** [review]
Looks good!

> Hmm... It looks like the tests are broken on `master`. Something with the new H.264 code. Open a new issue and ping the author(s) of that code?

I'll open a new issue.

**Pull request merged**